source/solvers/heat_transfer_assemblers.cc

```diff
@@ -0,0 +1,607 @@
+  const double dt  = time_steps_vector[0];
+  const double sdt = 1. / dt;
+
+  // Copy data elements
+  auto &strong_jacobian_vec = copy_data.strong_jacobian;
+  auto &local_matrix        = copy_data.local_matrix;
+
+  // assembling local matrix and right hand side
+  for (unsigned int q = 0; q < n_q_points; ++q)
+    {
+      if (simulation_parameters.multiphysics.free_surface)
+        {
+          // Calculation of the equivalent physical properties at the
+          // quadrature point
+          density =
+            calculate_point_property(scratch_data.phase_values[q],
+                                     physical_properties.fluids[0].density,
+                                     physical_properties.fluids[1].density);
+
+          specific_heat = calculate_point_property(
+            scratch_data.phase_values[q],
+            physical_properties.fluids[0].specific_heat,
+            physical_properties.fluids[1].specific_heat);
+
+          thermal_conductivity = calculate_point_property(
+            scratch_data.phase_values[q],
+            physical_properties.fluids[0].thermal_conductivity,
+            physical_properties.fluids[1].thermal_conductivity);
+
+          // Useful definitions
+          rho_cp = density * specific_heat;
+          alpha  = thermal_conductivity / rho_cp;
+        }
+
+      const auto method = this->simulation_control->get_assembly_method();
+
+      // Store JxW in local variable for faster access
+      const double JxW = JxW_vec[q];
+
+      const auto velocity = scratch_data.velocity_values[q];
+
+      // Calculation of the magnitude of the velocity for the
+      // stabilization parameter
+      const double u_mag = std::max(velocity.norm(), 1e-12);
+
+      // Calculation of the GLS stabilization parameter. The
+      // stabilization parameter used is different if the simulation is
+      // steady or unsteady. In the unsteady case it includes the value
+      // of the time-step
+      const double tau =
+        is_steady(method) ?
+          1. / std::sqrt(std::pow(2. * rho_cp * u_mag / h, 2) +
+                         9 * std::pow(4 * alpha / (h * h), 2)) :
+          1. /
+            std::sqrt(std::pow(sdt, 2) + std::pow(2. * rho_cp * u_mag / h, 2) +
+                      9 * std::pow(4 * alpha / (h * h), 2));
+
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        {
+          const auto phi_T_i      = scratch_data.phi_T[q][i];
+          const auto grad_phi_T_i = scratch_data.grad_phi_T[q][i];
+
+
+          for (unsigned int j = 0; j < n_dofs; ++j)
+            {
+              const Tensor<1, dim> grad_phi_T_j = scratch_data.grad_phi_T[q][j];
+
+
+              // Weak form for : - k * laplacian T + rho * cp *
+              //                  u * gradT - f -
+              //                  tau:grad(u) =0
+              // Hypothesis : incompressible newtonian fluid
+              // so tau:grad(u) =
+              // mu*(grad(u)+transpose(grad(u)).transpose(grad(u))
```

Comment:
Viscous dissipation should not be in the core assembler (and it is not right now no?)